vunk-parser/src/expr.rs

```diff
@@ -0,0 +1,585 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use chumsky::prelude::Rich;
+use chumsky::primitive::just;
+use chumsky::select;
+use chumsky::IterParser;
+use chumsky::Parser;
+
+use chumsky::span::SimpleSpan;
+use vunk_lexer::Token;
+
+use crate::literal::Float;
+use crate::literal::Integer;
+use crate::op::BinaryOp;
+use crate::op::UnaryOp;
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub enum Expr<'src> {
+    Bool(bool),
+    Integer(Integer),
+    Float(Float),
+    Str(&'src str),
+    List(Vec<Expr<'src>>),
+
+    Ident(&'src str),
+
+    Unary {
+        op: UnaryOp,
+        expr: Box<Expr<'src>>,
+    },
+
+    Binary {
+        lhs: Box<Expr<'src>>,
+        op: BinaryOp,
+        rhs: Box<Expr<'src>>,
+    },
+
+    LetIn {
+        exprs: Vec<Expr<'src>>,
+        sub: Box<Expr<'src>>,
+    },
+
+    IfElse {
+        condition: Box<Expr<'src>>,
+        tru: Box<Expr<'src>>,
+        fals: Box<Expr<'src>>,
+    },
+
+    TypeDef {
+        name: &'src str,
+        whereclause: Option<Vec<Generic<'src>>>,
+        members: Vec<Expr<'src>>,
+    },
+
+    EnumDef {
+        name: &'src str,
+        variants: Vec<Expr<'src>>,
+        whereclause: Option<Vec<Generic<'src>>>,
+    },
+
+    Decl {
+        ident: &'src str,
+        decl_type: DeclType<'src>,
+        generics: Option<Vec<&'src str>>,
+        whereclause: Option<Vec<Generic<'src>>>,
+    },
+
+    Def {
+        lhs: &'src str,
+        rhs: DefRhs<'src>,
+    },
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub enum DeclType<'src> {
+    TypeName(TypeName<'src>),
+    Func {
+        args: Vec<DeclArg<'src>>,
+        retty: TypeName<'src>,
+    },
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct TypeName<'src> {
+    pub name: &'src str,
+    pub generics: Option<Vec<&'src str>>,
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct Generic<'src> {
+    pub type_name: &'src str,
+    pub where_clause: Clauses<'src>,
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct Clauses<'src>(pub Vec<TypeName<'src>>);
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct DeclArg<'src> {
+    pub name: Option<&'src str>,
+    pub ty: DeclType<'src>,
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct DefRhs<'src> {
+    pub args: Vec<(&'src str, DeclType<'src>)>,
+    pub expr: Box<Expr<'src>>,
+}
+
+type ParserInput<'tokens, 'src> =
+    chumsky::input::SpannedInput<Token<'src>, SimpleSpan, &'tokens [(Token<'src>, SimpleSpan)]>;
+
+impl Expr<'_> {
+    pub fn parser<'tokens, 'src: 'tokens>() -> impl Parser<
+        'tokens,
+        ParserInput<'tokens, 'src>,
+        Expr<'src>,
+        chumsky::extra::Err<Rich<'tokens, Token<'src>, SimpleSpan>>,
+    > + Clone {
```

Comment:
I notice you have return types that look like this a lot. You can use either a trait alias (on nightly) or a trait + blanket impl to simplify these cases:

```rs
trait MyParser<'tokens, 'src, O> = Parser<
    'tokens,
    ParserInput<'tokens, 'src>,
    O,
    chumsky::extra::Err<Rich<'tokens, Token<'src>, SimpleSpan>>,
> + Clone;
```

or


```rs
trait MyParser<'tokens, 'src, O>: Parser<
    'tokens,
    ParserInput<'tokens, 'src>,
    O,
    chumsky::extra::Err<Rich<'tokens, Token<'src>, SimpleSpan>>,
> + Clone {}

impl<'tokens, 'src, P> MyParser<'tokens, 'src, O> for P
where
    P: Parser<
        'tokens,
        ParserInput<'tokens, 'src>,
        O,
        chumsky::extra::Err<Rich<'tokens, Token<'src>, SimpleSpan>>,
    > + Clone
{}
```

Then either case can be used on functions like

```rs
pub fn parser<'tokens, 'src: 'tokens>() -> impl MyParser<'tokens, 'src, Expr<'src>> {
    ...
}
```